File: Dotnet/DotNet.Comm/ClientPorts/AppPort/AppPort.cs
using DotNet.Utils.Controls.Utils;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DotNet.Comm.ClientPorts.AppPort
{
    /// <summary>
    /// Application Port 기준
    /// </summary>
    /// <remarks>
    /// Application ↔ OS 통신 Port
    /// </remarks>
    public abstract class AppPort : QYBindingBase
    {
        /// <summary>
        /// 통신 Port Log
        /// </summary>
        /// <remarks>
        /// 이 이벤트가 있는 Class는 통신 Port가 아닌 Application Port임<br/>
        /// Param[0] = Log 내용
        /// </remarks>
        public event Utils.Controls.Utils.Update_WithParam ComPortLog;

        private PortType _commType = PortType.Serial;
        private OSPort.OSPortBase _osPort = new OSPort.QYSerialPort();
        private bool _isAppOpen = false;

        /// <summary>
        /// OS Port 종류
        /// </summary>
        /// <remarks>
        /// 기존 종류와 다를경우 통신Port 신규 생성
        /// </remarks>
        public PortType PortType
        {
            get => this._commType;
            set
            {
                if(this._commType != value)
                {
                    if (value == PortType.Serial)
                        this._osPort = new OSPort.QYSerialPort();
                    else if (value == PortType.Ethernet)
                        this._osPort = new OSPort.QYEthernet(false);
                    this._osPort.Log += (msg) => { this.ComPortLog?.Invoke(msg); };

                    this._commType = value;

                    base.OnPropertyChanged(nameof(this.PortType));
                }
            }
        }
        /// <summary>
        /// OS ↔ Deivce 통신 Port
        /// </summary>
        public OSPort.OSPortBase OSPort { get => this._osPort; }
        /// <summary>
        /// Application Port Open여부
        /// </summary>
        public bool IsAppOpen
        {
            get => this._isAppOpen;
            protected set
            {
                if(this.IsAppOpen != value)
                {
                    this._isAppOpen = value;

                    base.OnPropertyChanged(nameof(this.IsAppOpen));
                }
            }
        }
        /// <summary>
        /// Port 연결
        /// </summary>
        /// <returns>연결 결과</returns>
        public abstract bool Connect();
        /// <summary>
        /// Port 연결 해제
        /// </summary>
        /// <returns>해제 결과</returns>
        public abstract bool Disconnect();
        /// <summary>
        /// OS Port에서 Data 읽기
        /// </summary>
        /// <returns>읽은 Data</returns>
        public abstract byte[] Read();
        /// /// <summary>
        /// OS Port에 Data 쓰기
        /// </summary>
        /// <param name="bytes">전송할 Data</param>
        /// <returns>실행 결과</returns>
        public abstract void Write(byte[] bytes);
        public abstract void Initialize();
    }
}

File: Flutter/qyapp/lib/conn/model/appport/appport.dart
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:qyapp/conn/model/osport/osport.dart';
import 'package:qyapp/conn/model/osport/serialport/android_serialport.dart';
import 'package:qyapp/conn/model/osport/serialport/desktop_serialport.dart';

///Application Port 기본틀
abstract class AppPortBase {
  //1. Fields
  int _osType = PortType.serial;
  OSPortBase? _osPort;

  //2. Property
  ///OS Port 종류<br/>
  ///기본값 Serial
  int get type => _osType;
  set type(int value){
    if(_osType != value){

      //OS Port 생성
      if(_osType == PortType.serial){
        if(Platform.isWindows || Platform.isIOS || Platform.isLinux){
          _osPort = DesktopSerialPort();
        }
        else if(Platform.isAndroid){
          _osPort = AndroidSerialport();
        }
      }

      _osType = value;
    }
  }
  ///OS Port
  OSPortBase get osPort => _osPort!;
  ///Application Port 열림 상태
  bool get isAppOpen;

  //3. 생성자
  ///Application Port 기본틀
  AppPortBase(){
    _init();
  }

  //4. Method - private
  ///초기 상태값 지정
  void _init(){
    if(type == PortType.serial){
      type = PortType.ethernet;
    }
    type = PortType.serial;
  }

  //4. Method - public
  ///Application 연결
  Future<bool> connect();
  ///Application 연결해제
  Future<bool> disconnect();
  ///Port Data 읽기
  Uint8List? read();
  ///Port Data 쓰기
  void write(Uint8List bytes);
  ///Port 초기화
  void initialize();
}